File: 2023/adventofcode/aoc.py
import abc
import re


def read_lines(file_path: str) -> list[str]:
    with open(file_path, 'r') as file:
        return [line.strip() for line in file]


def input_for_day(day: int) -> list[str]:
    return read_lines(f'data/day{day:02}.txt')


def parse_numbers(line: str) -> list[int]:
    return [int(number) for number in re.findall(r'-?\d+', line)]


class Solution(abc.ABC):
    def __init__(self, data: list[str]) -> None:
        self.data: list[str] = data

    @abc.abstractmethod
    def part1(self) -> int:
        ...

    @abc.abstractmethod
    def part2(self) -> int:
        ...

    def report(self):
        print(f'Part1: {self.part1()}')
        print(f'Part2: {self.part2()}')


def report(solution_class: type[Solution], day: int, __name__: str):
    if __name__ == '__main__':
        data = input_for_day(day)
        print(f'Day {day}')
        solution_class(data).report()

File: 2019/aoc.rb
# def read_input(day, separator = ';', strip: true)
#   lines = File.read("data/day#{String(day).rjust(2, '0')}.txt").split separator
#   lines = lines.collect(&:strip) if strip
#   lines
# end

# Class used to expose solutions and to cache intermediate values
class Solution
  def initialize(separator: "\n", strip: true)
    @data = load_data(day, separator, strip)
  end

  def day
    Integer(self.class.name[-2..])
  end

  def part1; end

  def part2; end

  def report
    puts "Day #{day}"
    puts "Part 1: #{part1}; Part 2: #{part2}"
  end

  def self.run
    new.report
  end

  def load_data(day, separator, strip)
    process_data(read_input(day, separator, strip))
  end

  def read_input(day, separator, strip)
    lines = File.read("data/day#{String(day).rjust(2, '0')}.txt").split(separator).filter
    lines = lines.collect(&:strip) if strip
    lines
  end

  def process_data(data)
    data
  end
end
